src/librustc_typeck/check/method/mod.rs

```diff
@@ -394,6 +435,84 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
     }
 }
 
+fn all_traits<'a>(ccx: &'a CrateCtxt) -> AllTraits<'a> {
+    if ccx.all_traits.borrow().is_none() {
+        use syntax::visit;
+
+        let mut traits = vec![];
+
+        // Crate-local:
+        //
+        // meh.
+        struct Visitor<'a> {
+            traits: &'a mut Vec<ast::DefId>
+        }
+        impl<'v,'a> visit::Visitor<'v> for Visitor<'a> {
+            fn visit_item(&mut self, i: &'v ast::Item) {
+                match i.node {
+                    ast::ItemTrait(..) => self.traits.push(ast_util::local_def(i.id)),
+                    _ => {}
+                }
+                visit::walk_item(self, i)
+            }
+        }
+        visit::walk_crate(&mut Visitor { traits: &mut traits }, ccx.tcx.map.krate());
+
+        // Cross-crate:
+        fn handle_external_def(traits: &mut Vec<ast::DefId>,
+                               cstore: &cstore::CStore,
+                               dl: decoder::DefLike,
+                               vis: ast::Visibility) {
+            // TODO: this seems to get reexports backwards (prints the
+            // path to the non-reexport, even if only the `pub use`
+            // reexport is actually public)
+            if vis != ast::Public { return }
+            match dl {
+                decoder::DlDef(def::DefTrait(did)) => traits.push(did),
+                decoder::DlDef(def::DefMod(did)) => {
+                    csearch::each_child_of_item(cstore, did, |dl, _, v| {
+                        handle_external_def(traits, cstore, dl, v)
+                    })
+                }
+                _ => {}
+            }
+        }
+        let cstore = &ccx.tcx.sess.cstore;
+        cstore.iter_crate_data(|&mut: cnum, _| {
+            csearch::each_top_level_item_of_crate(cstore, cnum, |dl, _, v| {
+                handle_external_def(&mut traits, cstore, dl, v)
+            })
+        });
+
+        *ccx.all_traits.borrow_mut() = Some(traits);
+    }
+
+    let borrow = ccx.all_traits.borrow();
+    assert!(borrow.is_some());
+    AllTraits {
+        borrow: borrow,
+        idx: 0
+    }
+}
+
+struct AllTraits<'a> {
+    borrow: cell::Ref<'a Option<Vec<ast::DefId>>>,
+    idx: usize
+}
+
+impl<'a> Iterator for AllTraits<'a> {
```

Comment:
Oh, bonus points for going all out. I would have just cloned the Vec and returned it ;)